nemo/collections/nlp/models/language_modeling/megatron_base_model.py

```diff
@@ -836,3 +842,37 @@ def _val_iterator_done(self, iterator):
             return iterator, True
         # reinsert the element back to the iterator
         return itertools.chain([element], iterator), False
+
+    def _get_max_steps(self):
+        """
+        Compute the maximum number of training steps.
+        """
+        if self._trainer.max_steps >= 0:
+            # Note that when `trainer.max_steps` is defined, we ignore `max_epochs` (even if training may end
+            # before `max_steps` is reached due to `max_epochs`). This is for backward compatibility with older
+            # versions of NeMo.
+            if self._trainer.max_epochs is not None and self._trainer.max_epochs >= 0:
+                logging.warning(
+                    "Ignoring `trainer.max_epochs` when computing `max_steps` because `trainer.max_steps` is already "
+                    f"set to {self._trainer.max_steps}."
+                )
+            return self._trainer.max_steps
+
+        if self._trainer.max_epochs is None or self._trainer.max_epochs < 0:
+            raise ValueError(
+                "Cannot compute `max_steps` if neither `trainer.max_steps` nor `trainer.max_epochs` is set"
+            )
+
+        # The number of training step per epoch is typically the number of global batches in the training set...
+        assert self._train_dl is not None
```

Comment:
Don't use asserts everywhere, check if trainer is none or not and fall back. It's not guaranteed to have trainer when building a non LLM model, but setup optimization and gets called on model restore when there is no trainer. So all these asserts raise the possibility of breaking model restore in untested cases.

Reply:
Ah thanks for pointing this out (though just a note that this is in `MegatronBaseModel` in nlp collection, so this shouldn't affect non-LLM models, right?)

I pushed a tentative fix in 4c991f553. Will mark this PR as draft while I'm testing it.

Reply:
> I pushed a tentative fix https://github.com/NVIDIA/NeMo/commit/3399ce842046c18e5852a075f1b457e426a8ff52. Will mark this PR as draft while I'm testing it.

Testing it made me realize I didn't remember correctly the code flow when writing this fix, and now I remember why I was triggering exceptions in `_get_max_steps()` if it couldn't be computed: this is because if we don't set `max_steps` properly, we fall back to the original issue where it's computed incorrectly.

One reason for my confusion (which I'm mentioning to be sure it doesn't confuse you as well) is that the `prepare_lr_scheduler()` call in `MegatronBaseModel.configure_optimizers()` ([here](https://github.com/NVIDIA/NeMo/blob/fd8f234070b21c64191824bf6b0f3d0cbdc2b61b/nemo/collections/nlp/models/language_modeling/megatron_base_model.py#L531)) is not actually the "usual" place where the scheduler is configured (this code is guarded by an `if` block) -- usually this happens in `ModelPT.setup_optimization()` ([here](https://github.com/NVIDIA/NeMo/blob/fd8f234070b21c64191824bf6b0f3d0cbdc2b61b/nemo/core/classes/modelPT.py#L733)).

So if we don't crash in `_get_max_steps()` (when we can't compute it), since we don't usually go through the check I added in `MegatronBaseModel.configure_optimizers()`, then we end up using a wrong schedule (same issue that this PR is trying to fix).

I pushed a new version in a26ac7535. This one actually fully relaxes the constraint that `max_steps` must be computed, and just skips the scheduler if it can't. This seems a bit dangerous to me though and I wanted to ask you if you think it'd be fine to trigger an exception in `MegatronBaseModel.configure_optimizers()` (not `setup_optimization()`), or if we also need that one to work even when the model is not fully setup for training?

Reply:
Silent about scheduler - it was a (bad) default in NeMo but it's easily detected by looking at logs.
For the megatron case, fine to assert for nlp only, other domains let them continue on without scheduler if max steps cannot be calculated (old default)

Reply:
> For the megatron case, fine to assert for nlp only, other domains let them continue on without scheduler if max steps cannot be calculated (old default)

Sounds good, I added a check in b6354f120 to raise an exception when the scheduler can't be instantiated in Megatron.